### PR TITLE
Health & Error Callback Implementation , Improved Abs Control Module 

### PR DESCRIPTION
**Pull request by yashrajsapra**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

**Description**
-> Added Health & Error Callbacks 
-> Improved Abs Control Module
-> Added Simple Control Module 

**Alternative(s) considered**

Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**

Type Choose one: (**Bug fix** | **Feature** | Documentation | Testing | Other)

**Screenshots (if applicable)**

**Checklist**

- [x] I have read the [Contribution Guidelines](https://github.com/Apra-Labs/ApraPipes/wiki/Contribution-Guidelines)
- [x] I have written Unit Tests
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach

### CODE REVIEW

File: base/test/ImageEncodeCV_tests.cpp [hunk 1]
```diff
@@ -14,6 +14,7 @@
 #include "PipeLine.h"
 #include "StatSink.h"
 #include "AbsControlModule.h"
```
Comment (mraduldubey) [inline base/test/ImageEncodeCV_tests.cpp:16]: dont need both headers here - only simple control module

**Review by mraduldubey — CHANGES_REQUESTED** [review]

File: base/include/TextOverlayXForm.h [hunk 2]
```diff
@@ -46,7 +46,6 @@ class TextOverlayXForm : public Module
 	bool validateInputPins();
 	bool validateOutputPins();
 	void addInputPin(framemetadata_sp &metadata, string &pinId);
-	void setProps(TextOverlayXForm);
```
Comment (mraduldubey) [inline base/include/TextOverlayXForm.h:49]: what is the idea here

File: base/src/ErrorObject.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                         const std::string &modName, const std::string &modId)
+    : mErrorCode(errCode), mErrorMessage(errorMsg), mModuleName(modName),
+      mModuleId(modId) {
+  mTimestamp = getCurrentTimestamp();
+}
+
+int ErrorObject::getErrorCode() const { return mErrorCode; }
+
+std::string ErrorObject::getCurrentTimestamp() const {
+  // Implement a function to get the current timestamp
```
[the remark this answers is not in the document]
Comment (mraduldubey) [reply base/src/ErrorObject.cpp:16]: seems incomplete

File: base/src/ErrorObject.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
+#include "ErrorObject.h"
+#include "Logger.h"
+#include <ctime>
+#include <iostream>
+
+ErrorObject::ErrorObject(int errCode, const std::string &errorMsg,
```
Comment (mraduldubey) [inline base/src/ErrorObject.cpp:6]: rename to APErrorObject

File: base/src/ErrorObject.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+void ErrorObject::setErrorCode(int errCode) {
+  mErrorCode = errCode;
+  mTimestamp = getCurrentTimestamp();
+}
+
+void ErrorObject::setErrorMessage(const std::string &errorMsg) {
+  mErrorMessage = errorMsg;
+  mTimestamp = getCurrentTimestamp();
```
Comment (mraduldubey) [inline base/src/ErrorObject.cpp:42]: if both setErrorCode and setErrorMessage are called which timestamp should be set ?

File: base/src/HealthObject.cpp [hunk 6]
```diff
@@ -0,0 +1,20 @@
+#include "HealthObject.h"
+#include "Logger.h"
```
Comment (mraduldubey) [inline base/src/HealthObject.cpp:2]: not required both: iostream and logger

File: base/src/HealthObject.cpp [hunk 7]
```diff
@@ -0,0 +1,20 @@
+#include "HealthObject.h"
```
Comment (mraduldubey) [inline base/src/HealthObject.cpp:1]: rename to APHealthObject

File: base/src/ImageEncoderCV.cpp [hunk 8]
```diff
@@ -56,15 +56,25 @@ class ImageEncoderCV::Detail
 
 	}
 
+	// void executeCallback() {
```
Comment (mraduldubey) [inline base/src/ImageEncoderCV.cpp:59]: clean

File: base/src/ImageEncoderCV.cpp [hunk 9]
```diff
@@ -56,15 +56,25 @@ class ImageEncoderCV::Detail
...
+	// 	}
+	// }
+
 	framemetadata_sp mOutputMetadata;
 	std::string mOutputPinId;
 	cv::Mat iImg;
 	vector<int> flags;
-
+	ErrorCallback mErrorCallback;
+    std::string mModuleId;
```
Comment (mraduldubey) [inline base/src/ImageEncoderCV.cpp:73]: why are these id and name requried here

File: base/src/ImageEncoderCV.cpp [hunk 10]
```diff
@@ -142,8 +152,10 @@ bool ImageEncoderCV::process(frame_container &frames)
 	{
 		return true;
 	}
+	// mDetail->executeCallback();
```
Comment (mraduldubey) [inline base/src/ImageEncoderCV.cpp:155]: clean

File: base/src/ImageEncoderCV.cpp [hunk 11]
```diff
@@ -159,3 +171,10 @@ bool ImageEncoderCV::processSOS(frame_sp &frame)
 	mDetail->setMetadata(metadata);
 	return true;
 }
+
+void ImageEncoderCV::registerErrorCallback(ErrorCallback callback)
+{
+	mDetail->mModuleId = getId();
+	mDetail->mModuleName = getName();
+	mDetail->mErrorCallback = callback;
```
Comment (mraduldubey) [inline base/src/ImageEncoderCV.cpp:179]: cool

File: base/src/SimpleControlModule.cpp [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+	return Module::sendEOS(frame);
+}
+
+void SimpleControlModule::sendEOPFrame()
+{
+	return Module::sendEoPFrame();
+}
+
+// Right Now, Just Logging But Can be used to Do bunch of other things 
+void SimpleControlModule::handleError(const ErrorObject &error)
```
Comment (mraduldubey) [inline base/src/SimpleControlModule.cpp:19]: i hope this method is virtual

File: base/test/ImageEncodeCV_tests.cpp [hunk 13]
```diff
@@ -175,7 +180,11 @@ BOOST_AUTO_TEST_CASE(RGB_profile, *boost::unit_test::disabled())
 
 	auto p = boost::shared_ptr<PipeLine>(new PipeLine("test"));
 	p->appendModule(fileReader);
+	p->addControlModule(mControl);
```
Comment (mraduldubey) [inline base/test/ImageEncodeCV_tests.cpp:183]: we need a step based test

**Review by mraduldubey — CHANGES_REQUESTED** [review]
go through comments, rename errorobject and healhobject - they are too generic

File: base/CMakeLists.txt [hunk 14]
```diff
@@ -195,6 +195,9 @@ SET(CORE_FILES
 	src/MotionVectorExtractor.cpp
 	src/OverlayModule.cpp
 	src/OrderedCacheOfFiles.cpp
+	src/SimpleControlModule.cpp
+	src/ErrorObject.cpp
```
[the remark this answers is not in the document]
Comment (yashrajsapra) [reply base/CMakeLists.txt:199]: Renamed to APErrorObject and APHealthObject

File: base/include/AbsControlModule.h [hunk 15]
```diff
@@ -1,40 +1,45 @@
 #pragma once
+#include "Callback.h"
```
[the remark this answers is not in the document]
Comment (yashrajsapra) [reply base/include/AbsControlModule.h:2]: Renamed to APCallback

File: base/include/Module.h [hunk 16]
```diff
@@ -188,9 +209,9 @@ class Module {
 	virtual bool produce() { return false; }
 	bool stepNonSource(frame_container& frames);
 	bool preProcessNonSource(frame_container& frames);
+	bool preProcessControl(frame_container& frames);
 	bool isRunning() { return mRunning; }
 
-	ModuleProps getProps();
```
[the remark this answers is not in the document]
Comment (yashrajsapra) [reply base/include/Module.h:193]: I made it Public, it was protected before

File: base/include/Module.h [hunk 17]
```diff
@@ -19,6 +19,7 @@
 #include "FIndexStrategy.h"
 #include "Command.h"
 #include "BufferMaker.h"
+#include "Callback.h"
```
[the remark this answers is not in the document]
Comment (yashrajsapra) [reply base/include/Module.h:22]: Done

File: base/include/TextOverlayXForm.h (hunk 2 above)
Comment (yashrajsapra) [reply base/include/TextOverlayXForm.h:49]: It was just a redundant code, We have declared but it's never defined

File: base/src/ErrorObject.cpp (hunk 4 above)
Comment (yashrajsapra) [reply base/src/ErrorObject.cpp:6]: Done

File: base/src/HealthObject.cpp (hunk 6 above)
Comment (yashrajsapra) [reply base/src/HealthObject.cpp:2]: Removed Iostream

File: base/src/HealthObject.cpp (hunk 7 above)
Comment (yashrajsapra) [reply base/src/HealthObject.cpp:1]: Done

File: base/src/SimpleControlModule.cpp (hunk 12 above)
Comment (yashrajsapra) [reply base/src/SimpleControlModule.cpp:19]: Yes it's a virtual function, defined in AbsControlModule

File: base/src/ImageEncoderCV.cpp (hunk 9 above)
Comment (yashrajsapra) [reply base/src/ImageEncoderCV.cpp:73]: To Propogate this information to Error Callback

File: base/src/ErrorObject.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (yashrajsapra) [reply base/src/ErrorObject.cpp:16]: Added

File: base/src/ErrorObject.cpp (hunk 5 above)
Comment (yashrajsapra) [reply base/src/ErrorObject.cpp:42]: Removed from both the places, it will be set while creating object

File: base/src/ImageEncoderCV.cpp (hunk 10 above)
Comment (yashrajsapra) [reply base/src/ImageEncoderCV.cpp:155]: Cleaned, Added Error Callback in GtkGlRenderer module

File: base/src/ImageEncoderCV.cpp (hunk 8 above)
Comment (yashrajsapra) [reply base/src/ImageEncoderCV.cpp]: Done

File: base/src/AbsControlModule.cpp [hunk 18]
```diff
@@ -8,78 +8,87 @@
 class AbsControlModule::Detail
 {
 public:
-    Detail(AbsControlModuleProps& _props) : mProps(_props)
-    {
-    }
+	Detail(AbsControlModuleProps& _props) : mProps(_props)
```
Comment (mraduldubey) [inline base/src/AbsControlModule.cpp:11]: add pragma once at top

**Review by mraduldubey — CHANGES_REQUESTED** [review]

File: base/src/APHealthObject.cpp [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+APHealthObject::APHealthObject(const std::string &modId) : mModuleId(modId)
+{
+  mTimestamp = getCurrentTimestamp();
+}
+
+std::string APHealthObject::getCurrentTimestamp() const
+{
+  // Implement a function to get the current timestamp
+  // For now, returning a placeholder
+  return "2024-07-12 10:00:00";
```
[the remark this answers is not in the document]
Comment (yashrajsapra) [reply base/src/APHealthObject.cpp]: Done, Added definition

**Review by mraduldubey — APPROVED** [review]

**Pull request merged**